### PR TITLE
66 create warning window

### PR DESCRIPTION
**Pull request by pavelparfishov**
<html>
<body>
<h2 dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px; font-size: 1.5em; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; padding-bottom: 0.3em; border-bottom: 1px solid var(--color-border-muted); color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Task merge request</h2>

Related Issue | [Создать окно предупреждений #66](https://github.com/SUAI-TaskPlanner-Contest/TaskPlanner/issues/66)
-- | --
Квалификация ревьюера | Знание требований заказчика

Исполнители: @pavelparfishov

<h3 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"></h3><h3 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a id="user-content-code-changes" class="anchor" aria-hidden="true" href="https://github.com/SUAI-TaskPlanner-Contest/TaskPlanner/blob/main/Workflow%20Guides/Templates/MrMessage.md#code-changes" style="box-sizing: border-box; background-color: transparent; color: var(--color-accent-fg); text-decoration: none; float: left; padding-right: 4px; margin-left: -20px; line-height: 1;"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg>
</a>Code changes</h3> Был написан QML-код для универсального окна ошибок.  Также были написаны конструкции на Python для вывода текста ошибки в окно. Информация об ошибках хранится в словаре. Еще были добавлены обработчики нажатия на кнопки и изменение цвета кнопки при нажатии для визуального удобства.
<h3 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 16px; font-size: 1.25em; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a id="user-content-ui-demonstration" class="anchor" aria-hidden="true" href="https://github.com/SUAI-TaskPlanner-Contest/TaskPlanner/blob/main/Workflow%20Guides/Templates/MrMessage.md#ui-demonstration" style="box-sizing: border-box; background-color: transparent; color: var(--color-accent-fg); text-decoration: none; float: left; padding-right: 4px; margin-left: -20px; line-height: 1;"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg>
</a>UI Demonstration</h3>

image

<h3 dir="auto" style="box-sizing: border-box; margin-top: 24px; margin-bottom: 0px !important; font-size: 1.25em; font-weight: var(--base-text-weight-semibold, 600); line-height: 1.25; color: rgb(230, 237, 243); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, &quot;Noto Sans&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a id="user-content-how-to-test" class="anchor" aria-hidden="true" href="https://github.com/SUAI-TaskPlanner-Contest/TaskPlanner/blob/main/Workflow%20Guides/Templates/MrMessage.md#how-to-test" style="box-sizing: border-box; background-color: transparent; color: var(--color-accent-fg); text-decoration: none; float: left; padding-right: 4px; margin-left: -20px; line-height: 1;"><svg class="octicon octicon-link" viewBox="0 0 16 16" version="1.1" width="16" height="16" aria-hidden="true"><path d="m7.775 3.275 1.25-1.25a3.5 3.5 0 1 1 4.95 4.95l-2.5 2.5a3.5 3.5 0 0 1-4.95 0 .751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018 1.998 1.998 0 0 0 2.83 0l2.5-2.5a2.002 2.002 0 0 0-2.83-2.83l-1.25 1.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042Zm-4.69 9.64a1.998 1.998 0 0 0 2.83 0l1.25-1.25a.751.751 0 0 1 1.042.018.751.751 0 0 1 .018 1.042l-1.25 1.25a3.5 3.5 0 1 1-4.95-4.95l2.5-2.5a3.5 3.5 0 0 1 4.95 0 .751.751 0 0 1-.018 1.042.751.751 0 0 1-1.042.018 1.998 1.998 0 0 0-2.83 0l-2.5 2.5a1.998 1.998 0 0 0 0 2.83Z"></path></svg>
</a>How to test</h3> <p>Проверить наличие необходимых полей и визуальную привлекательность окна. Также проверить корректность конструкций на Python</p>

</body>
</html>

### CODE REVIEW

**Comment by pavelparfishov** [comment]
@astronik00 Привет. Добавил обработку нажатий на кнопки и изменение цвета кнопок при нажатии и наведении курсора на кнопки.

**Comment by astronik00** [comment]
Убрать системное окно:
[https://evileg.com/ru/post/280/](https://vk.com/away.php?to=https%3A%2F%2Fevileg.com%2Fru%2Fpost%2F280%2F&cc_key=)

**Comment by astronik00** [comment]
> Вы удаляете задачу, у которой есть подзадачи! 

В принципе по действию это и так понятно. Смысл в том, что необходимо предупредить об удалении всех подзадач. Например: "У этой задачи есть подзадачи. В случае ее удаления все подзадачи будут также удалены". Выбор: "продолжить" и "отмена"

> Вы выходите из окна редактирования при наличии несохраненных данных!

Я думаю лучше: "В случае закрытия этого окна, все несохраненные данные будут утеряны" и тоже поменять на "продолжить" и "отмена"

> Вы удаляете сервер, у которого есть задачи!

"У этого сервера есть задачи. В случае удаления сервера все задачи будут также удалены". Выбор: "продолжить" и "отмена"

в общем поменьше экспрессии

**Comment by pavelparfishov** [comment]
@astronik00 Убрал системное окно. Изменил тексты.

**Comment by astronik00** [comment]
Отлично. Давай теперь шрифты одинаковые сделаем.
Попробуй шрифт Inter, я тебе скину позже, как он подключается. Сейчас узнаю это у Ярослава как раз.

**Comment by pavelparfishov** [comment]
@astronik00 Поменял шрифты, цвета. Добавил больше закруглений:

image

**Comment by astronik00** [comment]
> @astronik00 Поменял шрифты, цвета. Добавил больше закруглений:
> 
> image
> 

Отлично выглядит. Думаю кнопки надо побольше в высоту сделать и можно мерджить

**Comment by pavelparfishov** [comment]
@astronik00 Увеличил кнопки

image

File: Code/Warnings output/main.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    def __init__(self, error_info: Tuple[str, str, str, str]) -> None:
+        super().__init__(error_info[1])
+        self.error_info = ErrorInfo(*error_info)
+
+
+if __name__ == "__main__":
+    try:
+        a = 1
+        if a == 1:
+            raise CustomException(tuple(errors["ServerDeleteError"].values()))
```
Comment (astronik00) [inline Code/Warnings output/main.py:81]: Это из мейна нужно убрать. Создай документ markdown, в котором перенес бы этот код и показал, как вызвать твое окно.

File: Code/Warnings output/main.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+class ButtonHandler(QObject):
+    @pyqtSlot()
+    # Нажатие на левую кнопку
+    def button1Clicked(self):
+        print("Button 1 clicked")
+
+    # Нажатие на правую кнопку
+    @pyqtSlot()
+    def button2Clicked(self):
+        print("Button 2 clicked")
```
Comment (astronik00) [inline Code/Warnings output/main.py:68]: Здесь необходимо убрать print, лучше заменить пока на pass

File: Code/Warnings output/main.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        error_info = ErrorInfo(e.error_info.name, e.error_info.text, e.error_info.button1, e.error_info.button2)
+        context = engine.rootContext()
+        context.setContextProperty("errorInfo", error_info)
+
+        engine.load("window.qml")
+
+        if not engine.rootObjects():
+            sys.exit(-1)
+
+        sys.exit(app.exec_())
```
Comment (astronik00) [inline Code/Warnings output/main.py:98]: Соответственно, весь код, который вызывает твою ошибку из мейна, тоже нужно убрать. Оставить лишь вызов главного окна, который тут и лежал

Comment (pavelparfishov) [reply Code/Warnings output/main.py]: @astronik00 Почистил код. Вывел классы в отдельный файл. Сделал squash коммитов

**Pull request merged**